pkg/strcase/strcase.go

```diff
@@ -0,0 +1,70 @@
+package strcase
+
+import (
+	"strings"
+	"unicode"
+)
+
+var delimiters = map[rune]any{' ': struct{}{}, '_': struct{}{}, '-': struct{}{}, '.': struct{}{}}
+
+// Delimited converts a string to delimited.lower.case, here using `.` as delimiter.
+func Delimited(s string, d rune) string {
+	return convert(s, unicode.LowerCase, d)
+}
+
+// ScreamingDelimited converts a string to DELIMITED.UPPER.CASE, here using `.` as delimiter.
+func ScreamingDelimited(s string, d rune) string {
+	return convert(s, unicode.UpperCase, d)
+}
+
+// Snake converts a string to snake_case.
+func Snake(s string) string {
+	return Delimited(s, '_')
+}
+
+// ScreamingSnake converts a string to SCREAMING_SNAKE_CASE.
+func ScreamingSnake(s string) string {
+	return ScreamingDelimited(s, '_')
+}
+
+// convert converts a camelCase or space/underscore/hyphen/dot delimited string into various cases.
+// _case must be unicode.Lower or unicode.Upper.
+func convert(s string, _case int, d rune) string {
+	if s == "" {
+		return s
+	}
+
+	var wasLower bool
+	var wasNumber bool
+
+	s = strings.TrimSpace(s)
+
+	n := strings.Builder{}
+	n.Grow(len(s) + 2) // Allow adding at least 2 delimiters without another allocation.
+
+	for _, r := range s {
+		var isNumber bool
+		delimiter, isDelimiter := delimiters[r]
+		if !isDelimiter {
+			isNumber = unicode.IsNumber(r)
+			if wasNumber {
+				if !isNumber {
+					n.WriteRune(d)
+				}
+			} else if isNumber || wasLower && unicode.IsUpper(r) {
+				n.WriteRune(d)
+			}
+
+			n.WriteRune(unicode.To(_case, r))
+		} else if delimiter != d {
+			n.WriteRune(d)
+		}
```

Comment:
I'd just go for a `case ' ', '_', '-', '.':` instead of the `delimiters` map here.

Also, the value type of that map is `any`, thus the local variable `delimiter` is of type any and is either `nil` or `struct{}{}`. `d` on the other hand is of type `rune`, so `else if delimiter != d` will never be true.

Reply:
> I'd just go for a `case ' ', '_', '-', '.':` instead of the `delimiters` map here.

That's much simpler, yes. But I had in mind that at some point it might be necessary for the caller to extend the delimiter list. The current implementation would be prepared for this (at least a little bit) but I'll simplify it.

> ..., so `else if delimiter != d` will never be true.

It will never be `false`. This code path overrides considered delimiters with the desired one. So this should be just `else` or the code for `case ' ', '_', '-', '.':` in the simplified version.

Reply:
> It will never be `false`.

Err yes, it was `!=`.

> But I had in mind that at some point it might be necessary for the caller to extend the delimiter list.

The implementation is fixed to letters and numbers anyways, could simply treating everything else as a separator be an option?